### PR TITLE
Optimize `StringBuilder.Insert`.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
The character array overload calls the ReadOnlySpan overload, and the overloads of primitives call a private helper function that takes advantage of ISpanFormattable.

Salvaged from #58907.

### CODE REVIEW

**Review by stephentoub — COMMENTED** [review]
Thanks.  The changes overall look good, but there are some test failures to be looked at around corner case error handling.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -1290,17 +1290,24 @@ public StringBuilder Insert(int index, string? value)
...
+        public StringBuilder Insert(int index, byte value) => InsertSpanFormattable(index, value);
 
-        public StringBuilder Insert(int index, short value) => Insert(index, value.ToString(), 1);
+        public StringBuilder Insert(int index, short value) => InsertSpanFormattable(index, value);
 
         public StringBuilder Insert(int index, char value)
         {
+            if ((uint)index > (uint)Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), SR.ArgumentOutOfRange_Index);
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1308]: Is the throw worth moving into a throw helper?

Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1311]: Better for another time; throw helpers are very rarely used in `StringBuilder`.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Review by tannergooding — APPROVED** [review]

**Pull request merged**